tests/github_client/mod.rs

```diff
@@ -0,0 +1,876 @@
+//! `GithubClient` tests.
+//!
+//! These tests exercise the behavior of `GithubClient`. They involve setting
+//! up HTTP servers, creating a `GithubClient` to connect to those servers,
+//! executing some action, and validating the result.
+//!
+//! The [`run_test`] function is used to set up the test and give you a
+//! [`GithubClient`] to perform some action and validate its result.
+//!
+//! To write one of these tests, you'll need to use the recording function
+//! against the live GitHub site to fetch what the actual JSON objects should
+//! look like. To write a test, follow these steps:
+//!
+//! 1. Choose a repo in your GitHub account to run tests against. I use either
+//!    `ehuss/rust` or `ehuss/triagebot-test`, but any in your account will do.
+//!
+//!    **WARNING**: Do not write tests that modify the rust-lang org repos
+//!    like rust-lang/rust. Write those tests against your own fork (like
+//!    `ehuss/rust`). We don't want to pollute the real repos with things like
+//!    test PRs.
+//!
+//! 2. Make sure your `.env` file is configured with a `GITHUB_API_TOKEN` that
+//!    can access the repo you chose. See `README.md` for more about setting
+//!    this up.
+//!
+//! 3. Create a `#[test]` function following the form of the other tests below
+//!    where inside the `run_test` callback, execute the function you want to
+//!    exercise.
+//!
+//! 4. Run just a single test with recording enabled:
+//!
+//!    ```sh
+//!    TRIAGEBOT_TEST_RECORD_DIR=github_client/TEST_NAME_HERE cargo test \
+//!        --test testsuite -- --exact github_client::TEST_NAME_HERE
```

Comment:
How hard would it be to have a mode that runs the full test suite and checks/updates recordings against a repo that we control? e.g., rustbot/rust or similar?

My main worry about this design is that it gives us potentially false confidence if GitHub's APIs change, and it's somewhat painful to actually update when they do (potentially lots of local setup) vs. something that ideally we could run in CI, perhaps post-merge. Especially if we expect that new tests are primarily authored by "authorized" folks, we could have the CI in question be kickable with e.g. bors try or similar.

Reply:
I can take a look at making it easier to re-bless all of the tests using a single repo.

One concern is that if you just blanket re-record all of the tests, that will cause a very large diff since almost all of the data will change.